### ISSUE TITLE
[ENH] Add augmentation support to `TslibDataModule`

### ISSUE THREAD

**Issue by echo-xiao** [jdb78/pytorch-forecasting#2417]
Split out of #2330, where @phoeenniixx suggested opening a dedicated issue to
settle the design.

### What `thuml` does

`utils/augmentation.py` defines 14 transforms, in two groups:

- **Per-sample** (8): `jitter`, `scaling`, `rotation`, `permutation`,
  `magnitude_warp`, `time_warp`, `window_slice`, `window_warp`.
- **Cross-sample / label-guided** (6): `spawner`, `wdba`,
  `random_guided_warp(_shape)`, `discriminative_guided_warp(_shape)`. These
  need class labels and other samples, so they are classification-oriented and
  arguably out of scope for a forecasting data module.

They are applied through `run_augmentation_single`, gated on two conditions:

```python
# Dataset_ETT_hour / Dataset_ETT_minute / Dataset_Custom, in __read_data__
if self.set_type == 0 and self.args.augmentation_ratio > 0:
    self.data_x, self.data_y, tags = run_augmentation_single(
        self.data_x, self.data_y, self.args)
```

```python
# UEAloader, in __getitem__
if self.flag == "TRAIN" and self.args.augmentation_ratio > 0:
    batch_x, labels, tags = run_augmentation_single(batch_x, labels, self.args)
```

Two properties worth carrying over: augmentation is **train-split only**, and
it is **off unless explicitly enabled**.

**Comment by echo-xiao** [comment]
I went through `thuml`'s augmentation code and had a look at how our data
module is put together. Here's what I'd suggest. Happy to be talked out of any
of it.

### What thuml actually has

`utils/augmentation.py` has 14 transforms. Eight of them only look at a single
sample:

| transform | what it does |
|---|---|
| `jitter(x, sigma=0.03)` | adds gaussian noise |
| `scaling(x, sigma=0.1)` | multiplies each dimension by a random factor |
| `rotation(x)` | flips signs and permutes feature dimensions |
| `permutation(x, max_segments=5)` | reorders segments along time |
| `magnitude_warp(x, sigma=0.2, knot=4)` | warps amplitude with a cubic spline |
| `time_warp(x, sigma=0.2, knot=4)` | warps the time axis with a cubic spline |
| `window_slice(x, reduce_ratio=0.9)` | slices a random window, interpolates back |
| `window_warp(x, window_ratio=0.1)` | rescales a random window, keeps the edges |

The other six (`spawner`, `wdba`, the two `random_guided_warp` variants and the
two `discriminative_guided_warp` ones) need class labels and a pool of
same-class samples to warp towards.

They all go through `run_augmentation_single`, and it's always gated on two
things:

```python
# Dataset_ETT_hour / Dataset_ETT_minute / Dataset_Custom, in __read_data__
if self.set_type == 0 and self.args.augmentation_ratio > 0:
    self.data_x, self.data_y, tags = run_augmentation_single(
        self.data_x, self.data_y, self.args)
```

```python
# UEAloader, in __getitem__
if self.flag == "TRAIN" and self.args.augmentation_ratio > 0:
    batch_x, labels, tags = run_augmentation_single(batch_x, labels, self.args)
```

Train split only, and off unless you turn it on. Both worth keeping.

Notice the two call sites are in different places though. The forecasting
datasets augment once in `__read_data__`, the classification one augments on
every access in `__getitem__`. That difference turns out to matter, see below.

### What actually needs deciding

You already gave the API shape, so what's left is the mechanics. These are the
ones I think we have to settle before writing anything:

| # | question | why it isn't obvious |
|---|---|---|
| 1 | which transforms to start with | six of the fourteen need class labels we don't have |
| 2 | what the classes look like and where they live | affects how they get discovered and reused |
| 3 | one transform or a list | affects the API |
| 4 | which layer applies it | thuml does it two different ways, and the difference is not cosmetic |
| 5 | whether to use Lightning's batch transfer hook | it's the "official" place for this, but it has strings attached |
| 6 | what gets augmented, inputs or targets too | getting this wrong breaks context/target alignment silently |
| 7 | who guarantees train-only | getting this wrong contaminates evaluation silently |

Plus two smaller ones at the end that I'd rather leave to you.

The rest of this is my suggestion for each, in that order.

### 1. Which transforms to start with

I'd start with the eight per-sample transforms and skip the label-guided ones.
They need class labels and same-class neighbours, and a forecasting data module
has neither. They'd make sense once there's a classification data module in v2.

### 2 and 3. What the classes look like, and composing them

Your strategy pattern idea, following how `ScalerStrategy` is already written
(a `_BaseObject` subclass with `_tags` carrying an `object_type`, so
`all_objects` can find it):

```python
class BaseAugmentation(_BaseObject):
    _tags = {
        "object_type": "augmentation",
        "requires_target": False,   # does it need y, or just x
    }

    def transform(self, x, y=None):
        """Return the augmented (x, y). Called once per training window."""
```

```python
dm = TslibDataModule(..., augmentation=Jitter(sigma=0.03))
dm = TslibDataModule(..., augmentation=[Jitter(), Scaling()])
```

A list is applied in order. Default stays `None`. I'd put these under
`pytorch_forecasting/data/augmentation/` rather than `data/encoders/`, since
they're not encoders and don't have the fit/transform/inverse contract.

### 4. Which layer applies it

I'd do it in `_TslibDataset.__getitem__`, per window.

thuml's forecasting datasets augment in `__read_data__`, so each series gets
perturbed once and then every window and every epoch reuses that same noise.
That's a much weaker regulariser than resampling on each access, and it's
telling that their own `UEAloader` does the opposite.

There's also a trap if we put it in `_preprocess_data`: #2313 adds a
preprocessing cache in `_build_dataset`, which would freeze the noise for the
whole run. The code would look completely fine while doing nothing useful.

### 5. Why not `on_after_batch_transfer`?

That was my first instinct too, since it's the obvious Lightning hook for this
and wouldn't require touching `_TslibDataset` at all. Three things put me off
it (we don't currently use either batch transfer hook anywhere in the repo, for
what it's worth):

Gating on `self.trainer.training` breaks when the data module isn't attached to
a Trainer, since `dm.trainer` is `None` there. You'd need a `None` check.

More annoying: all our data module tests poke at `dm.train_dataset[0]` directly
without ever going through a Trainer, so the augmentation branch would simply
never run under the current testing style. We'd have to spin up a Trainer just
to test it.

And the hook runs in the main process after the batch is already on the GPU.
`time_warp` and `magnitude_warp` fit cubic splines, which is CPU work, so
that'd stall the training loop. In `__getitem__` it happens in the dataloader
workers instead, in parallel and overlapped with GPU compute.

### 6. What gets augmented

The whole window, history and future together.

This matters more than it looks. `jitter` and `scaling` are harmless if you
only touch the history, but `time_warp`, `window_slice` and `permutation` all
move things along the time axis. Applying those to the history alone would
quietly break the alignment between the context and the target it's supposed to
predict, with nothing raising an error. thuml augments `data_x` and `data_y`
together for exactly this reason.

Transforms that genuinely only touch the inputs can set
`requires_target=False` and get `y=None`.

### 7. Keeping it to the train split

We get this for free. `setup()` already builds four separate `_TslibDataset`
instances, one per split, so we just hand the transform to the train one:

```python
self.train_dataset = _TslibDataset(..., augmentation=self._augmentation)
self.val_dataset   = _TslibDataset(..., augmentation=None)
```

No `if training` anywhere, so there's nothing to get wrong. That's nicer than
gating at runtime, and much nicer than leaving it to each transform, where a
mistake would silently contaminate evaluation with no signal at all.

### One small thing on randomness

The transforms should draw from torch's RNG rather than `numpy.random`, so that
`seed_everything` and Lightning's per-worker seeding both work. thuml uses
`numpy.random` directly, which doesn't reseed per dataloader worker, so every
worker can end up with identical noise.

### How I'd split the work

| PR | contents |
|---|---|
| 1 | `BaseAugmentation`, the `augmentation` parameter on `TslibDataModule`, the train-only wiring, plus `Jitter` and `Scaling` with tests |
| 2 | the remaining six transforms |
| 3 | the same parameter on `EncoderDecoderTimeSeriesDataModule`, once the shape of the first PR is settled |

That way the design is reviewable in one small PR before any bulk of transforms
lands.

### Two things I'd rather you decide

Should we carry over `augmentation_ratio`? In thuml it controls how many
augmented copies get added to the dataset, not just whether each sample is
perturbed, so it changes the dataset length and interacts with windowing. I'd
leave it out of the first pass.

Should augmentation run before or after scaling? thuml does it after. Keeping
that order means the transforms see standardised values, which makes default
`sigma` values comparable across datasets. I'd keep it.

**Comment by phoeenniixx** [comment]
> They need class labels and same-class neighbours, and a forecasting data module
has neither.

Can you please elaborate on this? How DM is lacking class labels? I thought we had `y`? What exactly is missing and we need? Will #2082 help here?

---
I agree with most of your points, but i still need some clarifications:
- If most of it is 'train' only then it can happen in `stage=="fit"`, right?
- Why do we need to do it in `__getitem__` and not in `preprocess_data` - where we augment the data once and use the augmented data directly?

**Comment by echo-xiao** [comment]
1. we do have y, but it isn't a class label. they forecasting y is a numeric sequene of future values, the guided transoforms need "which class does this sample belong to", then sample another member of that class to warp towards. #2082 handles categorical encoding, but it doesn't give us that class semantics or the neighbor pool.
2. it can live in setup("fit"), but the gate has to separate train from val. that ranch builds both datasets, so stage == "fit" alone doesn't guarantee train-only.
3. reasons as follows:
    - augumentation only works if the same window looks different each time the model sees it. one draw per series means every window cut from it. and every epoch replays the same noise, so you just get a different fixed datasets.
    - after #2313 lands, _build_dataset caches _preprocess_data per series for the whole run, so augmentation there would run exactly once and still read as correct.
